src/instructlab/cli/model/upload.py

```diff
@@ -0,0 +1,239 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Standard
+from pathlib import Path
+import abc
+import logging
+import os
+
+# Third Party
+from huggingface_hub import HfApi
+from huggingface_hub import logging as hf_logging
+import click
+
+# First Party
+from instructlab import clickext
+from instructlab.configuration import DEFAULTS
+from instructlab.defaults import DEFAULT_INDENT
+from instructlab.utils import is_model_gguf, is_model_safetensors
+
+logger = logging.getLogger(__name__)
+
+
+class ModelUploader(abc.ABC):
+    """Base class for a model uploading backend"""
+
+    def __init__(
+        self,
+        ctx,
+        model: str,
+        destination: str,
+        release: str,
+    ) -> None:
+        self.ctx = ctx
+        self.model = model
+        self.destination = destination
+        self.release = release
+
+    @abc.abstractmethod
+    def upload(self) -> None:
+        """Uploads specified local model and stores it into destination@release"""
+
+
+class HFModelUploader(ModelUploader):
+    """Class to handle uploading safetensors and GGUF models to Hugging Face"""
+
+    def __init__(
+        self,
+        ctx,
+        model: str,
+        destination: str,
+        release: str,
+        hf_token: str,
+    ) -> None:
+        super().__init__(
+            ctx=ctx,
+            model=model,
+            destination=destination,
+            release=release,
+        )
+        self.hf_token = (
+            hf_token.strip()
+        )  # Remove trailing whitespaces from the token if they exist
+        self.local_model_path = ""
+
+    def upload(self):
+        """
+        Upload specified model to Hugging Face
+        """
+
+        # HF token check
+        if len(self.hf_token) == 0:
+            raise ValueError(
+                """HF_TOKEN var needs to be set in your environment to upload HF Model.
+                Alternatively, the token can be passed with --hf-token flag.
+                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
+            )
+
+        click.echo(
+            f"Uploading model to Hugging Face:\n{DEFAULT_INDENT}Model: {self.model}\n{DEFAULT_INDENT}Destination: {self.destination}@{self.release}\n"
+        )
+
+        # set logger
+        if self.ctx.obj is not None:
```

Comment:
interesting, what logging is handled specifically by HF that isn't already printed to stdout?

Reply:
This is the same thing we are doing in `download`

Reply:
ok thanks